File: frontend/src/lib/listingCardImage.ts
/** Prefer interior photos over Temer marketing flyer graphics for card thumbnails. */

const DIM_SUFFIX = /-\d+x\d+\.(jpe?g|png|webp)$/i
const TEMPLATE_MARKETING = /\/t-[a-z0-9_-]+\.(jpe?g|png|webp)$/i
const NUMBERED_INTERIOR = /\/\d{4}\/\d{2}\/\d+(?:-\d+)?\.(jpe?g|png|webp)$/i

export const MIN_CARD_IMAGE_SCORE = 10

const INTERIOR_HINTS = ['livingroom', 'kitchen', 'bedroom', 'interior', 'modern', 'hall', 'dining']

export function scoreCardImageUrl(url: string): number {
  const u = url.toLowerCase()
  let score = 0
  if (u.includes('-min') || u.includes('min.jpg')) score -= 25
  if (u.includes('logo') || u.includes('banner')) score -= 20
  if (u.includes('ayat-lomiyad') || u.includes('lomiyad')) score -= 15
  if (u.includes('/photo_')) score -= 10
  if (TEMPLATE_MARKETING.test(u) || u.includes('cenetr') || u.includes('t-cenetr')) score -= 24
  if (u.includes('-ads') || u.includes('goldmall')) score -= 14
  if (/\/0\d{3}[-_]/.test(u)) score -= 10
  if (INTERIOR_HINTS.some((hint) => u.includes(hint))) score += 24
  if (NUMBERED_INTERIOR.test(u)) score += 18
  if (u.includes('-1280x') || u.includes('-1920x')) score += 4
  if (DIM_SUFFIX.test(u) && !u.includes('-1280')) score -= 4
  return score
}

export function isCardImageUsable(url: string | null | undefined): boolean {
  if (!url?.trim()) return false
  return scoreCardImageUrl(url) >= MIN_CARD_IMAGE_SCORE
}

export function pickBestCardImageUrl(urls: Array<string | null | undefined>): string | null {
  const candidates = urls.filter((u): u is string => Boolean(u?.trim()))
  if (candidates.length === 0) return null
  const best = candidates.reduce((a, b) => (scoreCardImageUrl(a) > scoreCardImageUrl(b) ? a : b))
  return isCardImageUsable(best) ? best : null
}

File: backend/app/data/listing_card_image.py
"""Pick listing images that work as browse-card thumbnails (not marketing flyers)."""

from __future__ import annotations

import re

_DIM_SUFFIX = re.compile(r"-\d+x\d+\.(jpe?g|png|webp)$", re.I)
_TEMPLATE_MARKETING = re.compile(r"/t-[a-z0-9_-]+\.(jpe?g|png|webp)$", re.I)
_NUMBERED_INTERIOR = re.compile(r"/\d{4}/\d{2}/\d+(?:-\d+)?\.(jpe?g|png|webp)$", re.I)

MIN_CARD_IMAGE_SCORE = 10

_INTERIOR_HINTS = (
    "livingroom",
    "kitchen",
    "bedroom",
    "interior",
    "modern",
    "hall",
    "dining",
)


def score_card_image_url(url: str) -> int:
    u = url.lower()
    score = 0
    if "-min" in u or "min.jpg" in u:
        score -= 25
    if "logo" in u or "banner" in u:
        score -= 20
    if "ayat-lomiyad" in u or "lomiyad" in u:
        score -= 15
    if "/photo_" in u:
        score -= 10
    if _TEMPLATE_MARKETING.search(u) or "cenetr" in u or "t-cenetr" in u:
        score -= 24
    if "-ads" in u or "goldmall" in u:
        score -= 14
    if re.search(r"/0\d{3}[-_]", u):
        score -= 10
    for hint in _INTERIOR_HINTS:
        if hint in u:
            score += 24
            break
    if _NUMBERED_INTERIOR.search(u):
        score += 18
    if "-1280x" in u or "-1920x" in u:
        score += 4
    if _DIM_SUFFIX.search(u) and "-1280" not in u:
        score -= 4
    return score


def is_card_image_usable(url: str | None) -> bool:
    if not url or not url.strip():
        return False
    return score_card_image_url(url) >= MIN_CARD_IMAGE_SCORE


def pick_best_card_image_url(urls: list[str]) -> str | None:
    if not urls:
        return None
    best = max(urls, key=score_card_image_url)
    if score_card_image_url(best) < MIN_CARD_IMAGE_SCORE:
        return None
    return best


def order_images_for_card(urls: list[str]) -> list[str]:
    if not urls:
        return []
    return sorted(urls, key=score_card_image_url, reverse=True)
